### PR TITLE
Grading + caching for affine algebra of torus invariants

### PR DESCRIPTION
**Pull request by Lax202**

### CODE REVIEW

File: experimental/InvariantTheory/src/TorusInvariantsFast.jl [hunk 1]
```diff
@@ -264,14 +262,9 @@ julia> fundamental_invariants(RT)
...
 """
-function fundamental_invariants(z::TorGrpInvRing)
-    if isdefined(z, :fundamental)
-        return z.fundamental
-    else
-        R = z.representation
-        z.fundamental = torus_invariants_fast(weights(R), polynomial_ring(z))
-        return z.fundamental
-    end
+@attr function fundamental_invariants(z::TorGrpInvRing)
```
Comment (lgoettgens) [inline experimental/InvariantTheory/src/TorusInvariantsFast.jl:265]:
```suggestion
@attr SomeType function fundamental_invariants(z::TorGrpInvRing)
```
Please add the return type here to make the attribute lookup type stable

File: experimental/InvariantTheory/src/TorusInvariantsFast.jl [hunk 2]
```diff
@@ -399,10 +392,14 @@ julia> affine_algebra(RT)
 (Quotient of multivariate polynomial ring by ideal (-t[1]*t[3] + t[2]^2), Hom: quotient of multivariate polynomial ring -> graded multivariate polynomial ring)
 ```
 """
-function affine_algebra(R::TorGrpInvRing)
+@attr function affine_algebra(R::TorGrpInvRing)
```
Comment (lgoettgens) [inline experimental/InvariantTheory/src/TorusInvariantsFast.jl:395]:
```suggestion
@attr SomeType function affine_algebra(R::TorGrpInvRing)
```
Here as well

File: experimental/InvariantTheory/src/TorusInvariantsFast.jl (hunk 1 above)
Comment (Lax202) [reply experimental/InvariantTheory/src/TorusInvariantsFast.jl:265]: I don't know what type to use here other than Vector{MPolyRingElem}  or Vector{MPolyDecRingElem} - which are not concrete types.

Comment (Lax202) [reply experimental/InvariantTheory/src/TorusInvariantsFast.jl:265]: Faced the same issue with the other file - InvariantTheory.jl

Comment (lgoettgens) [reply experimental/InvariantTheory/src/TorusInvariantsFast.jl:265]: That's true. In most other places, we would add the field type as a type parameter to the invariant ring, which would then allow something smarter here.
But I think this is out of scope for this PR. 

@fingolfin can you keep this in mind for some future refactoring?

Comment (joschmitt) [reply experimental/InvariantTheory/src/TorusInvariantsFast.jl:265]: Yes, one would have to add parameters to the type `TorGrpInvRing` to make this more precise. I can have a look; I wanted to rename the types anyway in #3442.

Comment (joschmitt) [reply experimental/InvariantTheory/src/TorusInvariantsFast.jl:265]: Sorry, I typed to slow. We can keep this here as is and I look into it in #3442.

**Review by joschmitt — APPROVED** [review]
Looks good, I try to handle the return type business later.

**Comment by joschmitt** [comment]
The doctest needs to be adjusted: The return type of `fundamental_invariants` is now more precise.

File: experimental/InvariantTheory/src/TorusInvariantsFast.jl (hunk 1 above)
Comment (fingolfin) [reply experimental/InvariantTheory/src/TorusInvariantsFast.jl:265]: @lgoettgens sure, I actually had pointed out the same issue on a prior PR. We'll get there eventually ;-)

**Pull request merged**